### PR TITLE
Develop and run Jellyfish on balenaOS in balenaCloud

### PR DESCRIPTION
**Pull request by ab77**
Quickstart: https://github.com/product-os/jellyfish/blob/ab77/jellyfish-on-balena/docs/developing/on-balena.markdown

- https://github.com/product-os/environment-production/pull/277
- https://github.com/product-os/secrets/pull/46

TBCs:
- simplify/refactor all docs (i.e. remove docker|compose references, etc.) for development on-balena only
- refactor defaults handling between Taskfiles(s), balena.yml, package.json, etc. and jellyfish-environment
- refactor apps into systemd services and confd runtime (support mDNS .local resolution)
- rewire Grafana/Prometheus metrics/logs

Operational:
- remove docker test from branch protection rules on repository
- scale up `jellyfish-compute-resources-AutoScalingGroupSpot-F3K16JXB3ZTM` AutoScaling group via CFN (min x2)
- (optionally) update AWS/EC2 instance size
- update DNS to point to new ALBs (IPv4/IPv6 via AWS CNAME)

```
{api,jel,livechat}.ly.fish => jelly-LoadB-TAFQRKGBN0E5-167937939.us-east-1.elb.amazonaws.com
```

- (optionally) scale down Kubernetes productOS nodegroup(s) to x2 (minimum)
- remove Jellyfish Kubernetes assets from AWS/EKS production cluster (production-eks)
- remove balenaCI (Concourse) Jellyfish pipelines/assets

Future:
- refactor/replace Postgres/S3 tests with DB migrations (remove test dependency on external S3 object store in staging AWS account)

Change-type: major

### CODE REVIEW

**Comment by joshbwlng** [comment]
@ab77 Added a commit that adds in the `/config/env` endpoint we talked about today: https://github.com/product-os/jellyfish/pull/6847/commits/1772eaf29aff8026f92266d875c67f4579517b78

**Comment by joshbwlng** [comment]
@ab77 I think you may have ran `npm install` with node v16, take a look at the diffs for `package-lock.json` files 😂

**Comment by joshbwlng** [comment]
Here's some notes I ran into this time around (on Arch):

- needed to install `yq`
  - two similar tools called `yq`, need this one: https://github.com/mikefarah/yq
- `zsh: closing brace expected` error occurred when executing `uuid=$(printf...` command
  - no error when ran in bash
- `git secret reveal -f` => `git-secret: abort: directory '.gitsecret' does not exist. Use 'git secret init' to initialize git-secret`
- had issues resolving [uuid].local
  - tried setting up avahi, but no luck yet, so temporarily hardcoded record in `/etc/hosts`
- `make push args_extra="--env MDNS_TLD=${tld} DNS_TLD=${tld}`:
```
Some services failed to build:
        jellyfish-tests: Get https://registry-1.docker.io/v2/resinci/jellyfish-test/manifests/v1.4.15: unauthorized: please use personal access token to login
        api: Get https://registry-1.docker.io/v2/balena/open-balena-base/manifests/v13.0.2: unauthorized: please use personal access token to login
        ui: Get https://registry-1.docker.io/v2/balena/open-balena-base/manifests/v13.0.2: unauthorized: please use personal access token to login
        livechat: Get https://registry-1.docker.io/v2/balena/open-balena-base/manifests/v13.0.2: unauthorized: please use personal access token to login
        postgres: (HTTP code 500) server error - Get https://registry-1.docker.io/v2/balena/open-balena-db/manifests/v4.1.0: unauthorized: please use personal access token to login
        redis: (HTTP code 500) server error - Get https://registry-1.docker.io/v2/library/redis/manifests/6-alpine: unauthorized: please use personal access token to login
        s3: (HTTP code 500) server error - Get https://registry-1.docker.io/v2/balena/open-balena-s3/manifests/build-ab77-jellyfish-on-balena: unauthorized: please use personal access token to login
        registry: (HTTP code 500) server error - Get https://registry-1.docker.io/v2/balena/open-balena-registry/manifests/v2.22.0: unauthorized: please use personal access token to login
        prometheus: Get https://registry-1.docker.io/v2/prom/prometheus/manifests/v2.31.1: unauthorized: please use personal access token to login
        grafana: Get https://registry-1.docker.io/v2/grafana/grafana/manifests/8.2.5: unauthorized: please use personal access token to login
        mdns: (HTTP code 500) server error - Get https://registry-1.docker.io/v2/balena/balena-mdns-publisher/manifests/v1.14.0: unauthorized: please use personal access token to login
        balena-ca: (HTTP code 500) server error - Get https://registry-1.docker.io/v2/balena/ca-private/manifests/v0.0.2: unauthorized: please use personal access token to login
        cert-manager: (HTTP code 500) server error - Get https://registry-1.docker.io/v2/balena/cert-manager/manifests/v0.0.8: unauthorized: please use personal access token to login
        haproxy: Get https://registry-1.docker.io/v2/balena/open-balena-haproxy/manifests/v4.0.5: unauthorized: please use personal access token to login
        tag-sidecar: (HTTP code 500) server error - Get https://registry-1.docker.io/v2/library/bash/manifests/alpine3.14: unauthorized: please use personal access token to login
        wifi-connect: Get https://registry-1.docker.io/v2/balenalib/intel-nuc-debian/manifests/latest: unauthorized: please use personal access token to login
```

**Comment by ab77** [comment]
> Here's some notes I ran into this time around (on Arch):

@joshbwlng we should talk through these in a meet, most are expected and either have workarounds or solutions.

**Comment by ab77** [comment]
> 

> @ab77 I think you may have ran `npm install` with node v16, take a look at the diffs for `package-lock.json` files 😂

Yep, this is an annoyance in part due to balena-cli and `${NODE_VERSION}` collisions. Should be fixed.

**Comment by ab77** [comment]
@apostolism @Stefsak folks, any change to get balenaCloud API tokens with access to the balenaCloud orgs set up as GH org secrets for all of our GH orgs (specifically in this case product-os) to unblock us from moving Jellyfish to balenaCloud?

Suggested approach:
1. create a new team https://dashboard.balena-cloud.com/orgs/196119/teams (e.g. names:  machines, automation, services, etc.)
2. create/add a new user to this team (e.g. github)
3. create an balenaCloud API token and create `BALENA_API_TOKEN` secret at the GitHub org level with its value
4. great new team full access to all fleets in the org
5. repeat for all the other orgs in balenaCloud

cc @joshbwlng @Ereski @LucianBuzzo @klutchell

**Comment by ab77** [comment]
@balena-ci retest

**Comment by ab77** [comment]
@LucianBuzzo @karaxuna @Ereski @joshbwlng @edufschmidt folks, if you have time to check, this PR contains a WIP revamped Livepush/Local Mode workflow against a balenaOS device. The [doc](https://github.com/product-os/jellyfish/blob/ab77/jellyfish-on-balena/docs/developing/on-balena.markdown).

**Comment by Ereski** [comment]
image

Reviewing this is gonna be a problem.

**Comment by ab77** [comment]
> image
> 
> Reviewing this is gonna be a problem.

No problem, that's what self-certify is for :)

**Comment by joshbwlng** [comment]
@Ereski FWIW I'm cherry picking useful bits from this branch that we can use now, but it will still probably be a large PR.

File: docs/developing/on-balena.markdown [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+tld=ly.fish.local
+
+mkdir -p balena
+
+uuid=$(printf "results:\n$(sudo balena scan)" \
+  | yq e '.results[] | select(.osVariant=="development").host' - \
+  | awk -F'.' '{print $1}') \
+  && balena_device_uuid=$(balena device ${uuid::7} | grep UUID | cut -c24-)
+```
```
Comment (joshbwlng) [inline docs/developing/on-balena.markdown:42]: This block assumes `bash` and doesn't work with `zsh`. We should probably move this to a script that works with both and make it callable from `npm run ...`.

File: docs/developing/on-balena.markdown [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+balena device local-mode ${uuid} --enable
+```
+
+* create `~/.balena/secrets.json` containing your DockerHub credentials
+
+```json
+{
+  "https://index.docker.io/v1/": {
+    "username": "{{username}}",
+    "password": "{{password}}"
```
Comment (joshbwlng) [inline docs/developing/on-balena.markdown:66]: This "password" actually needs to be an access token.

File: docs/developing/on-balena.markdown [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+```PowerShell
+Import-Certificate -FilePath "${NODE_EXTRA_CA_CERTS}" `
+  -CertStoreLocation Cert:\LocalMachine\Root
+```
+
+* configure operating system (**Arch Linux**)
+
+```sh
+sudo trust anchor --store pki/ca.pem
```
Comment (joshbwlng) [inline docs/developing/on-balena.markdown:115]:
```suggestion
sudo trust anchor --store ${NODE_EXTRA_CA_CERTS}
```

File: docs/developing/on-balena.markdown [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+sudo yum install ca-certificates \
+  && sudo update-ca-trust force-enable \
+  && sudo cp ${NODE_EXTRA_CA_CERTS} /etc/pki/ca-trust/source/anchors/productOS.pem \
+  && update-ca-trust extract
+```
+
+* verify that all of backends are up
+
+```sh
+open https://balena:${balena_device_uuid}@stats.${balena_device_uuid}.${tld}:1936/metrics
```
Comment (joshbwlng) [inline docs/developing/on-balena.markdown:137]: `open` isn't a universally available command, should probably generalize these.

File: docs/developing/on-balena.markdown [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+```
+
+* open Jellyfish
+
+```sh
+open https://jel.${balena_device_uuid}.${tld}
```
Comment (joshbwlng) [inline docs/developing/on-balena.markdown:143]: Was able to open this and `stats` in my browser, but the certificate isn't completely trusted for Chrome. No problems with Firefox though, so probably some missing steps for Chrome? (On Arch)

**Comment by joshbwlng** [comment]
@ab77 Went through the steps again today and got everything working! Left some comments that we'll have to get to at some point, but was able to get everything running and accessible. 👍

File: docs/developing/on-balena.markdown (hunk 1 above)
Comment (ab77) [reply docs/developing/on-balena.markdown:42]: it's really just a change to `... ${uuid:0:7}` which makes it work in both bash and zsh.

File: docs/developing/on-balena.markdown (hunk 5 above)
Comment (ab77) [reply docs/developing/on-balena.markdown:143]: Definitely should be working, try restarting the whole stack `systemctl restart balena` on the device. On the initial push, some cervices may still be running with stale config.

File: docs/developing/on-balena.markdown (hunk 4 above)
Comment (ab77) [reply docs/developing/on-balena.markdown:137]: Yeah, I know `open` is available on macOS and possibly Windows. Not sure about various Linux desktops though.

**Comment by ab77** [comment]
> * The initial push didn't work

With WSL2 + mDNS you are not going to have a great time with default settings. The WSL VM is going to be on its own network by default, which means it won't see any mDNS broadcasts from your LAN. It's somewhat documented [here](https://github.com/balena-io/balena-on-balena/blob/master/docs/development.md#fn4). You can read the lengthy discussion [here](https://github.com/microsoft/WSL/issues/4150) and see if you can somehow reconfigure your WSL v2 VM to run on the network bridge. Aside from that, you could see in Windows 11 changed this default behaviour or try using a different hypervisor (e.g. VirtualBox) from your Linux workstation. You could also try downgrading to WSL v1, since it uses host network by default (not NAT like v2) and may work.

> * The balena-ca and s3 images are massive (500+mb each)

This is a known issue/fact. The upstream images are what they are and we can try to pull them apart and optimise, with the usual maintenance responsibility transfer.

> * The build finished and started up, but I was unable to login because the UI was sending requests to the API at https://api.db0c73f094ae787294b9bcc2bfd90961.bob.local/ which doesn't resolve to anything. Is the `.bob` part of the address correct

bob.local shouldn't be anywhere in the docs/env-vars for this fleet/deployment, so this looks like either a symptom of the WSL issue or some left over settings from the previous deployment. See if you can sort out mDNS resolution first.

@LucianBuzzo

**Comment by ab77** [comment]
@balena-ci retest

**Comment by ab77** [comment]
@balena-ci retest

**Comment by ab77** [comment]
@taromurao it would be great if you could run through on-balena markdown with your NUC and let me know if you encounter any issues.

**Comment by taromurao** [comment]
@ab77 I don't have a NUC. Would you, @joshbwlng and @Ereski, test on NUC? I don't know at the moment who else are using NUC.

**Comment by ab77** [comment]
> @ab77 I don't have a NUC. Would you, @joshbwlng and @Ereski, test on NUC? I don't know at the moment who else are using NUC.

you should request one.

**Comment by ab77** [comment]
@balena-ci retest

**Comment by taromurao** [comment]
Let us know loudly when the PR gets merged!

**Comment by ab77** [comment]
> Let us know loudly when the PR gets merged!

Realistically we won't be able to merge this until the team moves to developing on balena.

**Comment by ab77** [comment]
@balena-ci retest

**Comment by joshbwlng** [comment]
@ab77 *Finally* making my way back to this, hopefully to help get it merged this time around. Got it up and running again. Had some mDNS issues that I seem to have atleast somewhat solved, but need to investigate why long form UUID mDNS isn't working for me (Arch). Also why Chrome/Chromium doesn't like the installed certs while Firefox works fine.

Do you happen to know if we can silence these logs? They confused me at first making me think something was wrong:
```
[Logs]    [3/16/2022, 5:57:24 PM] [haproxy-sidecar] + docker ps --filter 'name=haproxy_' --filter 'status=running' --filter 'network=1_default' --format '{{.ID}}'
[Logs]    [3/16/2022, 5:57:24 PM] [haproxy-sidecar] + '[' d26f45b56cf1 '='  ]
[Logs]    [3/16/2022, 5:57:24 PM] [haproxy-sidecar] + docker ps --filter 'name=haproxy_' --filter 'status=running' --filter 'network=1_default' --format '{{.ID}}'
[Logs]    [3/16/2022, 5:57:24 PM] [haproxy-sidecar] + haproxy=d26f45b56cf1
[Logs]    [3/16/2022, 5:57:24 PM] [haproxy-sidecar] + '[' d26f45b56cf1 '=' d26f45b56cf1 ]
[Logs]    [3/16/2022, 5:57:24 PM] [haproxy-sidecar] + sleep 20s
[Logs]    [3/16/2022, 5:57:26 PM] [postgres] 2022-03-16 08:57:26.079 UTC [812] FATAL:  database "docker" does not exist
```

That last one is apparently harmless even though it has the big scary F word in it. :sweat_smile: Haven't investigated yet, but I'm guessing it has something to do with the `postgres` service's healthcheck test: `pg_isready -U "${POSTGRES_USER}"`.

Have some small fixes/updates that I will probably be able to push tomorrow to make the setup a bit less painful.

**Comment by ab77** [comment]
> Do you happen to know if we can silence these logs?

Yes, there is an implicit `set -x` on this service, I've made a change to only set it when `VERBOSE=true`.

**Comment by joshbwlng** [comment]
@ab77 I may have asked you this before, but what do you think about making the switch from k8s to balena in a couple phases instead of all at once in a single PR? Seems like we could deploy to both for a while and then update DNS records to point to balena instead of k8s when we think things are ready. If anything goes wrong, we could easily switch back to the same version running on k8s by reverting those DNS record changes.

**Comment by ab77** [comment]
> @ab77 I may have asked you this before, but what do you think about making the switch from k8s to balena in a couple phases instead of all at once in a single PR? Seems like we could deploy to both for a while and then update DNS records to point to balena instead of k8s when we think things are ready. If anything goes wrong, we could easily switch back to the same version running on k8s by reverting those DNS record changes.

It's not trivial given the amount of changes in the apps/composition. If you have an idea how to make both work concurrently, sharing the same code base, then sure, otherwise I don't think it's worth the pain.

**Comment by joshbwlng** [comment]
@ab77 That would be an interesting thing to take a look at: how can we break this up into more digestable and therefore less dangerous chunks so we can make this transition as safely as possible? Some things I think are obvious - dropping deploy templates last and probably bringing in some docker-compose services first. Those shouldn't break production and, if we're careful, *probably* shouldn't break development. I'll see if I can start breaking some of this up into smaller PRs over the next few days.

**Comment by ab77** [comment]
@balena-ci retest

**Comment by ab77** [comment]
@balena-ci retest

**Comment by ab77** [comment]
@joshbwlng I've removed balena API key-pair from secrets and enabled cert-manager to auto-generate these (for tests). 

For production, the keys are passed to balenaCloud API as `JF_WEBHOOK_SIGN_PRIVATE_KEY` and `JF_WEBHOOK_PUBLIC_KEY` and the API is told where JF lives via`JF_WEBHOOK_URI` (e.g. https://api.ly.fish/api/v2/hooks/balena-api).

**Comment by ab77** [comment]
@joshbwlng I am struggling with `catch-uncommitted` step:

```
[jellyfish-tests]  [catch-uncommitted] > catch-uncommitted --skip-node-versionbot-changes --exclude=VERSION,.versionbot/,.balena/,Dockerfile.template,.gitsecret/,.gitsecret~HEAD/,secrets/
[jellyfish-tests]  [catch-uncommitted] 
[jellyfish-tests]  [catch-uncommitted] 
[jellyfish-tests]  [catch-uncommitted] diff --git a/.gitsecret b/.gitsecret
[jellyfish-tests]  [catch-uncommitted] deleted file mode 120000
[jellyfish-tests]  [catch-uncommitted] index e3ad6fbad..000000000
[jellyfish-tests]  [catch-uncommitted] --- a/.gitsecret
[jellyfish-tests]  [catch-uncommitted] +++ /dev/null
[jellyfish-tests]  [catch-uncommitted] @@ -1 +0,0 @@
[jellyfish-tests]  [catch-uncommitted] -secrets/.gitsecret
[jellyfish-tests]  [catch-uncommitted] \ No newline at end of file
[jellyfish-tests]  [catch-uncommitted] diff --git a/.gitsecret~HEAD b/.gitsecret~HEAD
[jellyfish-tests]  [catch-uncommitted] deleted file mode 120000
[jellyfish-tests]  [catch-uncommitted] index e3ad6fbad..000000000
[jellyfish-tests]  [catch-uncommitted] @@ -1 +0,0 @@
[jellyfish-tests]  [catch-uncommitted] -secrets/.gitsecret
[jellyfish-tests]  [catch-uncommitted] ** Uncommitted changes found (^^^ diff above ^^^) - FAIL **
[jellyfish-tests]  task: Failed to run task "catch-uncommitted": exit status 1
```

Is this check even required? I am not sure what purpose is serves, since the code has obviously been committed...

**Comment by LucianBuzzo** [comment]
@ab77 Let's just delete the test. I don't think it's serving any purpose and is generating noise.

**Comment by joshbwlng** [comment]
@ab77 @LucianBuzzo The `catch-uncomitted` test checks that `npm run docs:diagram` and `npm run docs:summary` output is in sync with what was committed. These aren't mission critical, but it is nice to make sure the related diagram/docs are kept up-to-date. If we don't care about keeping them updated, then we should drop them altogether as well IMO. We *could* do this in a local commit hook or something, but it would be unnnecessary 99% of the time.

I also seem to remember that this uncommitted check has caught package-lock.json diffs in the past, for example when a developer pushed a package-lock.json generated by an older version of npm than what we expect in testing/production.

**Review by LucianBuzzo — APPROVED** [review]

**Comment by LucianBuzzo** [comment]
@balena-ci rebase

**Comment by ab77** [comment]
@cmfcruz @joshbwlng [jellyfish-compute-resources](https://us-east-1.console.aws.amazon.com/cloudformation/home?region=us-east-1#/stacks/stackinfo?filteringStatus=active&filteringText=jellyfish&viewNested=true&hideStacks=false&stackId=arn%3Aaws%3Acloudformation%3Aus-east-1%3A491725000532%3Astack%2Fjellyfish-compute-resources%2F2cb40560-ddd5-11eb-8b36-1260e6a1f297) currently running with x6 2xlarge instances (spot) in [jellyfish-on-balena](https://dashboard.balena-cloud.com/fleets/1842831/summary) fleet.

DNS has been updated to `jelly-LoadB-TAFQRKGBN0E5-167937939.us-east-1.elb.amazonaws.com` (old: `94ece80e-productos-jellyfi-973f-662494725.us-east-1.elb.amazonaws.com
`).

I've scaled down the pods in k8s for zero and dropped the product-os nodegroup to x3 instances (from x6).

Grafana dashboards will need to be adjusted to pick up metrics.

cc @LucianBuzzo

**Pull request merged**